instrumentation/src/main/java/io/opentelemetry/android/features/diskbuffering/SignalDiskExporter.kt

```diff
@@ -0,0 +1,155 @@
+import java.util.concurrent.TimeUnit
+
+/**
+ * Entrypoint to read and export previously cached signals.
+ */
+class SignalDiskExporter internal constructor(
+    private val spanDiskExporter: SpanDiskExporter?,
+    private val metricDiskExporter: MetricDiskExporter?,
+    private val logRecordDiskExporter: LogRecordDiskExporter?,
+    private val exportTimeoutInMillis: Long,
+) {
+    /**
+     * A batch contains all the signals that arrived in one call to [SpanDiskExporter.export]. So if
+     * that function is called 5 times, then there will be 5 batches in disk. This function reads
+     * and exports ONE batch every time is called.
+     *
+     * @return TRUE if it found data in disk and the exporter succeeded. FALSE if any of those conditions were
+     * not met.
+     */
+    @WorkerThread
+    @Throws(IOException::class)
+    fun exportBatchOfSpans(): Boolean {
+        return spanDiskExporter?.exportStoredBatch(
+            exportTimeoutInMillis,
+            TimeUnit.MILLISECONDS,
+        ) ?: false
+    }
+
+    /**
+     * A batch contains all the signals that arrived in one call to [MetricDiskExporter.export]. So if
+     * that function is called 5 times, then there will be 5 batches in disk. This function reads
+     * and exports ONE batch every time is called.
+     *
+     * @return TRUE if it found data in disk and the exporter succeeded. FALSE if any of those conditions were
+     * not met.
+     */
+    @WorkerThread
+    @Throws(IOException::class)
+    fun exportBatchOfMetrics(): Boolean {
+        return metricDiskExporter?.exportStoredBatch(
+            exportTimeoutInMillis,
+            TimeUnit.MILLISECONDS,
+        ) ?: false
+    }
+
+    /**
+     * A batch contains all the signals that arrived in one call to [LogRecordDiskExporter.export]. So if
+     * that function is called 5 times, then there will be 5 batches in disk. This function reads
+     * and exports ONE batch every time is called.
+     *
+     * @return TRUE if it found data in disk and the exporter succeeded. FALSE if any of those conditions were
+     * not met.
+     */
+    @WorkerThread
+    @Throws(IOException::class)
+    fun exportBatchOfLogs(): Boolean {
+        return logRecordDiskExporter?.exportStoredBatch(
+            exportTimeoutInMillis,
+            TimeUnit.MILLISECONDS,
+        ) ?: false
+    }
+
+    /**
+     * Convenience method that attempts to export all kinds of signals from disk.
+     *
+     * @return TRUE if at least one of the signals were successfully exported, FALSE if no signal
+     * of any kind was exported.
+     */
+    @WorkerThread
+    @Throws(IOException::class)
+    fun exportBatchOfEach(): Boolean {
+        var atLeastOneWorked = exportBatchOfSpans()
+        if (exportBatchOfMetrics()) {
+            atLeastOneWorked = true
+        }
+        if (exportBatchOfLogs()) {
+            atLeastOneWorked = true
+        }
+        return atLeastOneWorked
+    }
+
+    class Builder {
```

Comment:
Usually, Builders in Kotlin are a bit odd (and have downsides) since the language provides named and default parameters.
https://www.baeldung.com/kotlin/builder-pattern for context

Reply:
This seems related to [this comment](https://github.com/open-telemetry/opentelemetry-android/pull/221#discussion_r1454126269) from @breedx-splk, I haven't taken a deep look yet but probably what he suggests might avoid the need to use a builder altogether.

Reply:
Looking at `createInstance`, all you need is a data class with optional parameters, so you create the instance with all the parameters at once, if they are null or not it's not important, the final object is created at the very end and immutable, it's only important to construct all the parameters beforehand which you have to do with the builder anyway.

```kotlin
val instance =
            createInstance(
                spanDiskExporter,
                metricDiskExporter,
                logRecordDiskExporter,
                timeoutInMillis,
            )
```

becomes

```kotlin
val instance = SignalDiskExporter(
  spanDiskExporter = spanDiskExporter,
  metricDiskExporter = metricDiskExporter,
  ..
)
```

Reply:
If the Java OTel SDK has builders everywhere, maybe still using builders makes sense, so it's not different everywhere, your call here, just wanted to share the Builders is not a Kotlin-friendly pattern

Reply:
> Looking at `createInstance`, all you need is a data class with optional parameters, so you create the instance with all the parameters at once, if they are null or not it's not important, the final object is created at the very end and immutable, it's only important to construct all the parameters beforehand which you have to do with the builder anyway.

Are you referring to the function `createInstance` within `SignalDiskExporterTest`? Because if that's the case, I agree, a builder is unnecessary there. My concern is with `OpenTelemetryRumBuilder` where the span exporter is created in `buildSpanExporter` which is called from `buildTracerProvider`, which is called from `build` where the OpenTelemetry instance is created, and the setup of `SignalDiskExporter` is done right after. The way the OpenTelemetry instance is built is by setting each signal up independently, which involves a couple of steps, so for the spans one, we need to set a tracer provider with a span processor with the span exporter, and the last one in that chain (the exporter setup) is where we can attach a disk exporter. There are many functions involved in setting the span signal up, and I'm guessing the same will happen with the other 2 signals, so in this scenario is handy to have a place to collect each disk exporter until they're all ready to create the `SignalDiskExporter` instance. It doesn't necessarily need to be called a builder though, but it's essentially what it is so it made sense. I'm not sure if Kotlin has another name for it or a more Kotlin-y way to handle these cases. I'm thinking that maybe after merging [this PR](https://github.com/open-telemetry/opentelemetry-java-contrib/pull/1161) we might get other options to address this without a builder though.

Reply:
gotcha, yeah I was looking at the `createInstance` as an example, just that you could replicate the same approach but got your point now, thanks.
I don't know the OTel codebase at all, just getting used to it by doing reviews and skimming the code once in a while so feel free to ignore me when I tell nonsense :D

Reply:
No worries, I might miss a lot of things when creating these PRs too, especially recent features added to Kotlin so I appreciate any feedback, in this case, we could actually get rid of the builder is just that there might be different ways to create this instance and there's not a clear "best one" to me, plus it might change after this [PR gets](https://github.com/open-telemetry/opentelemetry-java-contrib/pull/1161) merged. I like these kinds of conversations, although I'm looking forward to be able to go through them in an Android SIG otherwise it's a lot of typing 😅

Reply:
Indeed. Haha